### PR TITLE
Testing for SSO feature

### LINKED ISSUE #155: Add tests for SSO feature
- Add e2e test to cover behaviour of SSO feature (test *operator* functionality, not the docker image)
- Add any relevant unit tests for added utility methods, controller statements, etc.

### PR DESCRIPTION
**Pull request by edavidj**
**What this PR does / why we need it?**:

- E2E Test that verifies:
  - Env Var values are pulled from the secret + SSO config in CRD
  - Updates those values at the pod level when the secret is updated
  - Removes values when SSO is disabled
  - Multiple OIDC/Oauth2 providers are processed
  - Multiple OIDC/Oauth2 providers are updated correctly
- Unit Tests
  - Verifies behaviour of the util.CustomizeEnvSSO

**Which issue(s) this PR fixes**:

Fixes #155

### CODE REVIEW

File: test/e2e/openliberty_sso.go [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+	if err != nil {
+		return err
+	}
+
+	secret.Data = map[string][]byte{
+		"github-clientId": []byte("different="),
+		"github-clientSecret": []byte("differentSecret="),
+		"oidc-clientId": []byte("bW9vb29vb28="),
+		"oidc-clientSecret": []byte("dGhlbGF1Z2hpbmdjb3c="),
+		"twitter-clientId": []byte("twitterID="),
```
Comment (leochr) [inline test/e2e/openliberty_sso.go:128]: Twitter is a bit different. It has consumerKey and consumerSecret.

File: test/e2e/openliberty_sso.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+	// TODO: Update secret and verify values
+	secret := corev1.Secret{}
+	err = f.Client.Get(goctx.TODO(), secretTarget, &secret)
+	if err != nil {
+		return err
+	}
+
+	secret.Data = map[string][]byte{
+		"github-clientId": []byte("different="),
```
Comment (leochr) [inline test/e2e/openliberty_sso.go:124]: Please use base64 encoded values

File: test/e2e/openliberty_sso.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+		return err
+	}
+
+	// Wait for old pod to finish deleting
+	err = util.WaitForPodUpdates(t, f, ctx, target, 1)
+	if err != nil {
+		return err
+	}
+
+	if err = verifyEnvVariables(ctx, f, target, secretTarget); err == nil {
```
Comment (leochr) [inline test/e2e/openliberty_sso.go:169]: Instead of this check, it'll be good to go through environment variables and make sure that there were no env variable that begins with SEC_SSO_

File: pkg/utils/utils_test.go [hunk 4]
```diff
@@ -85,10 +87,99 @@ func TestCustomizeLibertyEnv(t *testing.T) {
...
+	data :=  map[string][]byte{
+			"github-clientId": []byte("bW9vb29vb28="),
+			"github-clientSecret": []byte("dGhlbGF1Z2hpbmdjb3c="),
+			"oidc-clientId": []byte("bW9vb29vb28="),
+			"oidc-clientSecret": []byte("dGhlbGF1Z2hpbmdjb3c="),
+	}
+	pts := &corev1.PodTemplateSpec{}
+	ssoSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name+"-sso",
```
Comment (leochr) [inline pkg/utils/utils_test.go:139]: Should be name + "-olapp-sso"

File: pkg/utils/utils_test.go [hunk 5]
```diff
@@ -85,10 +87,99 @@ func TestCustomizeLibertyEnv(t *testing.T) {
...
+		{"Github clientid set", string(data["github-clientId"]), podEnv["SEC_SSO_GITHUB_CLIENTID"]},
+		{"Github clientSecret set", string(data["github-clientSecret"]), podEnv["SEC_SSO_GITHUB_CLIENTSECRET"]},
+		{"OIDC clientId set", string(data["oidc-clientId"]), podEnv["SEC_SSO_OIDC_CLIENTID"]},
+		{"OIDC clientSecret set", string(data["oidc-clientSecret"]), podEnv["SEC_SSO_OIDC_CLIENTSECRET"]},
+		{"Github hostname set", "github.com", podEnv["SEC_SSO_GITHUB_HOSTNAME"]},
+		{"OIDC discovery endpoint", "myapp.mycompany.com", podEnv["SEC_SSO_OIDC_DISCOVERYENDPOINT"]},
+		{"custom1 authorization endpoint", "specify-required-value", podEnv["SEC_SSO_CUSTOM1_AUTHORIZATIONENDPOINT"]},
+		{"custom1 token endpoint", "specify-required-value", podEnv["SEC_SSO_CUSTOM1_TOKENENDPOINT"]},
+		{"custom2 authorization endpoint", "specify-required-value", podEnv["SEC_SSO_CUSTOM2_AUTHORIZATIONENDPOINT"]},
+		{"custom2 token endpoint", "specify-required-value", podEnv["SEC_SSO_CUSTOM2_TOKENENDPOINT"]},
```
Comment (leochr) [inline pkg/utils/utils_test.go:161]: Each SSO parameter is processed individually. It'll be good to verify in unit tests that ALL applicable parameters for SSO are processed correctly.

File: test/e2e/openliberty_sso.go [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+	err = util.WaitForPodUpdates(t, f, ctx, target, 1)
+	if err != nil {
+		return err
+	}
+
+	if err = verifyEnvVariables(ctx, f, target, secretTarget); err != nil {
+		return err
+	}
+	t.Log("secret data updated successfully")
+
```
Comment (leochr) [inline test/e2e/openliberty_sso.go:147]: It'll be good to cover the scenario when the only update to secret is updating the value of an existing key (no new keys). It was an edge case for which we had to add some extra code (to force Deployment to pass the value to Pod).

**Review by leochr — COMMENTED** [review]
@edavidj Great job! Looks good. Some comments for enhancements. Thanks.

File: test/e2e/openliberty_sso.go (hunk 2 above)
Comment (edavidj) [reply test/e2e/openliberty_sso.go]: After some experimenting, this [post](https://github.com/kubernetes/client-go/issues/403#issuecomment-386760053) seems accurate in that the JSON encoding done by the go client encodes these as base64. So the test is more accurate with just the original strings when creating the Secret through the go client.

File: pkg/utils/utils_test.go [hunk 7]
```diff
@@ -85,10 +87,139 @@ func TestCustomizeLibertyEnv(t *testing.T) {
...
+		{Name: "SEC_TLS_TRUSTDEFAULTCERTS", Value: "true"},
+		{Name: "SEC_IMPORT_K8S_CERTS", Value: "true"},
+	}
+	spec.Expose = &expose
+	spec.Route = &v1beta1.OpenLibertyApplicationRoute{
+		Host:        "myapp.mycompany.com",
+		Termination: &terminationPolicy,
+		Certificate: &v1beta1.Certificate{},
+	}
+	spec.SSO = &v1beta1.OpenLibertyApplicationSSO{
```
Comment (leochr) [inline pkg/utils/utils_test.go:110]: Please add common parameters that apply to all providers: `sso.redirectToRPHostAndPort` and `sso.mapToUserRegistry`

**Review by leochr — APPROVED** [review]
@edavidj Looks good. Great job! Thank you. One minor comment about the unit test. Please merge once that's addressed.
